### PR TITLE
Fix problems with the unavailability of the sessions in the App::before

### LINKED ISSUE #282: [L5.5.] Locale picker component does not change language
##### Expected behavior
When I change language with `localePicker` component it should change language to my selection.

##### Actual behavior
Language is not changed and it is always set to English (default). The ajax request has no errors and it redirects back successfully.

##### Reproduce steps
Add `localePicker` to layout file. Change language using the component dropdown.

##### October build
Build 419 with Laravel 5.5

### LINKED ISSUE #283: [L5.5] A temporary solution for localePicker
`LocalePicker` still doesn't work correctly without "Force URL schema".

My solution:

1. Update `Plugin.php`,  `boot` function 

```
\Cms\Classes\CmsController::extend(function($controller) {
    $controller->middleware('RainLab\Translate\Middlewares\Locale');
});
```

2. New middleware in `RainLab\Translate\Middlwares\`

```
class Locale
{
    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next)
    {
        $translator = Translator::instance();
        $translator->isConfigured();
    
        if( !$translator->loadLocaleFromRequest() ) {
            $translator->loadLocaleFromSession();
        }
        
        return $next($request);
    }
}
```

3. Update `routes.php`

```
Event::listen('cms.beforeRoute', function() {
        
      foreach( \RainLab\Translate\Models\Locale::listAvailable() as $code => $locale ) {
            
            Route::middleware('web')->prefix($code)->group( function() {
                Route::any('{slug?}', 'Cms\Classes\CmsController@run')->where( 'slug', '(.*)?' );
            } );
            
      }
    
});
```
in `App::before` `Closure`

Tested on custom theme and `Bonjour`
And it works correctly.

I something broke?

build 420 with L5.5

### PR DESCRIPTION
**Pull request by hocopor**
Fixes #283, Fixes #282

### CODE REVIEW

File: middlewares/Locale.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (LukeTowers) [inline middlewares/Locale.php]: This folder should be named `middleware` instead of `middlewares`

File: middlewares/Locale.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        $translator = Translator::instance();
+        $translator->isConfigured();
+    
+        if( !$translator->loadLocaleFromRequest() ) {
```
Comment (LukeTowers) [inline middlewares/Locale.php]: Could you please add some comments (either directly in the code or just here on Github) explaining exactly what is happening here?
Is this just copied over from the App::before()? Why has a lot of the other logic there been removed?

File: routes.php [hunk 3]
```diff
@@ -7,44 +7,19 @@
...
-    }
-
-    /*
-     * Register routes
-     */
-    Route::group(['prefix' => $locale, 'middleware' => 'web'], function() {
-        Route::any('{slug}', 'Cms\Classes\CmsController@run')->where('slug', '(.*)?');
-    });
-
-    Route::any($locale, 'Cms\Classes\CmsController@run');
```
Comment (hocopor) [inline routes.php]: We dont need do it. I just change `{slug}` as optional `{slug?}`

File: middlewares/Locale.php (hunk 1 above)
Comment (hocopor) [inline middlewares/Locale.php]: done

File: middlewares/Locale.php (hunk 2 above)
Comment (hocopor) [inline middlewares/Locale.php]: That logic needed just for declare only one `Route`.
I don't care about it and create routes for all locales in the `cms.beforeRoute` event.

So i set needed locale with this code
```
 $translator = Translator::instance();

        if( !$translator->loadLocaleFromRequest() ) {
            $translator->loadLocaleFromSession();
        }
```
in `Global Middleware` — `RainLab\Translate\Middleware\Locale`

File: routes.php [hunk 4]
```diff
@@ -1,50 +1,24 @@
...
-
-    if (!$translator->loadLocaleFromRequest()) {
-        $translator->loadLocaleFromSession();
-        return;
-    }
-
-    if (!$locale = $translator->getLocale()) {
-        return;
-    }
-
```
Comment (hocopor) [inline routes.php]: This logic needed just for declare only one `Route`.
I don't care about it and create `Routes` for all locales in the `cms.beforeRoute` event.

So i set needed locale with this code

 ```
$translator = Translator::instance();

        if( !$translator->loadLocaleFromRequest() ) {
            $translator->loadLocaleFromSession();
        }
```
in `Global Middleware` — `RainLab\Translate\Middleware\Locale`

**Comment by LukeTowers** [comment]
@hocopor why'd you close this?

**Comment by hocopor** [comment]
@LukeTowers I want to clean up the repository but don't want to bother you with a notifications :)

**Comment by LukeTowers** [comment]
Alright, thanks. Resubmit this PR once you've got it rebased of the lastest master branch please :)

**Comment by manogi** [comment]
Hello there - is there a fix somewhere by now for this? I am an October 419 and Translate 1.2.11 and the my Session variables are gone in all routes that contain a locale...

**Comment by LukeTowers** [comment]
@manogi if you check the changelog on the marketplace you can see that newer versions are marked as pending, which means that you need to enable edgeUpdates in config/cms.php

**Comment by manogi** [comment]
@LukeTowers Ah, that is something I did not know, thank you for informing me!

**Comment by daftspunk** [comment]
Alt solution in 76b798e89d345957802888142a190e5643a5eafe should fix it.

**Pull request closed without merging**